pkg/health/checker/checker.go

```diff
@@ -361,224 +180,347 @@ func watchConsulHealth(
+type HealthClient interface {
+	HealthCheckEndpoints(ctx context.Context, req *hclient.HealthEndpointsRequest) (map[string]health.HealthState, error)
+	HealthMonitor(ctx context.Context, req *hclient.HealthRequest, respCh chan *hclient.HealthResponse) error
+}
+
+type healthChecker struct {
+	healthClient HealthClient
+	serviceNodes func(serviceID string) ([]types.NodeName, error)
+}
+
+func NewHealthChecker(
+	hClient HealthClient,
+	serviceNodes func(serviceID string) ([]types.NodeName, error),
+) HealthChecker {
+	return healthChecker{
+		healthClient: hClient,
+		serviceNodes: serviceNodes,
+	}
+}
+
+func (h healthChecker) WatchPodOnNode(
 	ctx context.Context,
-	serviceID string,
-	nodeIDs []types.NodeName,
-	nodeIDsCh <-chan []types.NodeName,
-	resultCh chan<- map[types.NodeName]health.Result,
-	errCh chan<- error,
-	watchDelay time.Duration,
-	useHealthService bool,
-	useOnlyHealthService bool,
+	nodeID types.NodeName,
+	podID types.PodID,
 	status manifest.StatusStanza,
-) {
-	// nodes are considered always healthy when status port is not set
-	// when always healthy, WatchService ONLY sends the healthy status repeatedly on a timer set to delay
-	alwaysHealthy := status.Port == 0
-	if h.useHealthService || useHealthService || useOnlyHealthService || h.useOnlyHealthService {
-		respChan := make(chan *hclient.HealthResponse, len(nodeIDs))
-		go func() {
-			oldResultsDelay := time.Second * 5
-			healthResults := make(map[types.NodeName]health.Result)
-			if alwaysHealthy {
-				// reduce the delay to send old results since
-				// when the nodes are always healthy, these
-				// are the only messages sent
-				oldResultsDelay = time.Second * 2
-				for _, nodeID := range nodeIDs {
-					healthResults[nodeID] = health.Result{
-						ID:      types.PodID(serviceID),
-						Node:    nodeID,
-						Service: serviceID,
-						Status:  health.Passing,
+) (chan health.Result, chan error) {
+	resultCh := make(chan health.Result)
+	errCh := make(chan error)
+
+	endpoint := nodeIDToStatusEndpoint(nodeID, status)
+	protocol := "HTTPS"
+	if status.HTTP {
+		protocol = "HTTP"
+	}
+	healthReq := &hclient.HealthRequest{
+		Url:      endpoint,
+		Protocol: protocol,
+	}
+	respCh := make(chan *hclient.HealthResponse, 1)
+	healthResult := health.Result{
+		ID:      podID,
+		Node:    nodeID,
+		Service: podID.String(),
+		Status:  health.Unknown,
+	}
+	tick := time.Tick(time.Second * 2)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case resp := <-respCh:
+				endpoint := resp.HealthRequest.Url
+				nodeID, err := statusURLToNodeName(endpoint)
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						return
+					case errCh <- err:
 					}
+					continue
+				}
+				healthResult = health.Result{
+					ID:      podID,
+					Node:    nodeID,
+					Service: podID.String(),
+					Status:  resp.Health,
+				}
+				select {
+				case <-ctx.Done():
+					return
+				case resultCh <- healthResult:
+				}
+			case <-tick:
```

Comment:
What's the intention of this case? Is it to provide a best effort view of health in case the health service is slow? Why is that important?

Reply:
To mimic the behavior of the old consul checker and also I think it's possible that messages are lost over the network. Health service only sends updates when there's a change.